python/cudf/cudf/dataframe/categorical.py

```diff
@@ -52,25 +51,49 @@ def set_categories(self, new_categories):
         col = self._set_categories(new_categories)
         return Series(data=col)
 
-    def _set_categories(self, new_categories):
+    def _set_categories(self, new_categories, is_unique=False):
         """Returns a new CategoricalColumn with the categories set to the
-        specified *new_categories*."""
-        if new_categories is self._categories:
+        specified *new_categories*.
+
+        Notes
+        -----
+        Assumes ``new_categories`` is the same dtype as the current categories
+        """
+        if new_categories is self.categories:
             return self._parent.copy()
-        codemap = {v: i for i, v in enumerate(new_categories)}
-        h_recoder = np.zeros(
-            len(self.categories), dtype=self._parent.data.dtype
-        )
-        for i, catval in enumerate(self.categories):
-            h_recoder[i] = codemap.get(catval, self._parent.default_na_value())
-        # recode the data buffer
-        recoded = cudautils.recode(
-            self._parent.data.to_gpu_array(),
-            h_recoder,
-            self._parent.default_na_value(),
-        )
-        buf_rec = Buffer(recoded)
-        return self._parent.replace(data=buf_rec, categories=new_categories)
+
+        from cudf import DataFrame
+
+        # Join the old and new categories to build a map from
+        # old to new codes, inserting na_sentinel for any old
+        # categories that don't exist in the new categories
+
+        new_cats = columnops.as_column(new_categories)
+
+        # Ensure new_categories is unique first
+        if not is_unique:
+            new_cats = new_cats.unique()
+
+        new = DataFrame({'cats': new_cats,
+                         'new': cudautils.arange(len(new_cats))})
+        old = DataFrame({'cats': self.categories,
+                         'old': cudautils.arange(len(self.categories))})
+
+        # Join the old and new categories and line up their codes
+        pos = old.merge(new, on='cats', how='left').drop('cats')
+        # Sort by the old codes so the new col is a map from
+        # old -> new code, e.g. codemap[old_code] = new_code
+        pos = pos.sort_values('old').reset_index(drop=True)
+
+        codes = self._parent.data.to_gpu_array()
+        na_sentinel = self._parent.default_na_value()
+        old_to_new_codes_map = pos['new'].to_gpu_array(fillna=na_sentinel)
```

Comment:
This chunk of code looks unideal in general, should this just use nulls instead of Pandas's `-1` for missing values and then handle that in the conversion code?

Reply:
I left the -1 logic in because that's what we were doing before, and I was trying to target this PR only at moving the categories from CPU -> GPU. You're right, we could use nulls to represent codes that exist in the old set of categories, but not the new one. In Arrow at least, the index values for null slots is undefined, so ending up with garbage values there should be fine.

We'd need a more advanced routine than the current `cudautils.recode()`, one which constructs the new null bitmap alongside the new codes buffer. I'd be happy to modify `recode` to do that in numba, but this feels like something that probably already exists/should exist in libcudf?

Reply:
👍 let's do this in a followup PR, I believe this can just be a left join which would construct the nulls you want anyway.